aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp

```diff
@@ -480,6 +481,22 @@ void block_sparse_mm(
       mat1.values().is_contiguous() ||
       mat1.values().transpose(-2, -1).is_contiguous());
 
+  // NOTE: the code below allows arbitrary block sizes
+  // and might be potentially faster than cuSPARSE implementation
+  // especially for not very sparse inputs.
+  if (mat1.scalar_type() == ScalarType::Half || mat1.scalar_type() == ScalarType::BFloat16) {
+    at::native::sparse::impl::_compressed_row_strided_addmm_out(
+        mat1, // ignored when beta == 0
+        mat1,
+        mat2,
+        /*beta=*/0.,
```

Comment:
We have to handle beta properly. 

The core function should always evaluate `C = \alpha (A@B) + \beta C`. If C is an explicit out parameter, or allocated automatically, and if beta == 0, alpha == 1 as for `mm` should be handled above this level. 

If you trace a call from `mm`, `mm_out`, `addmm` or `addmm_out` you should be able to see the allocation of self/result, as applicable at different levels and where this function is eventually called. 

The two existing cuda variants `at::cuda::sparse::bsrmm` and `at::cuda::sparse::spmm(` and the mkl variant `at::mkl::sparse::mm` all have a generally consistent call pattern. They all take as input:
 - Info defining a sparse matrix A
 - Info defining a dense matrix B
 - A scalar alpha applied to the intermediate result A@B
 - Result matrix C of compatible shape for A@B, which it is understood will be overwritten
 - A scalar \beta which will be applied to C before accumulating the result onto it

C should hold after calling `C = \beta C + \alpha (A @ B)

`torch.mm` is a special case where C is not provided, instead is pre-allocated as zero, beta is zero and \alpha is one.
`torch.addmm` is a special case where `self` is used to initialize C, but is not modified in-place.

Reply:
Good catch! Updating my code now...